dolphinscheduler-registry/dolphinscheduler-registry-plugins/dolphinscheduler-registry-raft/src/main/java/org/apache/dolphinscheduler/plugin/registry/raft/EphemeralNodeManager.java

```diff
@@ -0,0 +1,405 @@
+                    final String nodeAddress = entry.getKey();
+                    activeMasterServers.remove(nodeAddress);
+                    updateActiveMaster(activeMasterServers);
+                    addDeadServer(Constants.MASTER_TYPE, nodeAddress);
+                    if (nodeAddress.split(Constants.COLON)[0].equals(NetUtils.getHost())) {
+                        connectionState = ConnectionState.DISCONNECTED;
+                        triggerListener(ConnectionState.DISCONNECTED);
+                        removeNodeData(nodeAddress);
+                    }
+                    log.warn("Master server {} connect to raft cluster timeout, last heartbeat {}, timeout config {} ms",
+                            nodeAddress, convertTimeToString(entry.getValue()), expireTime);
+                }
+            }
+            Map<String, Long> activeWorkerServers = getActiveWorkerServers();
+            for (Map.Entry<String, Long> entry : activeWorkerServers.entrySet()) {
+                if ((System.currentTimeMillis() - entry.getValue()) > expireTime) {
+                    final String nodeAddress = entry.getKey();
+                    activeWorkerServers.remove(nodeAddress);
+                    updateActiveWorker(nodeAddress, activeWorkerServers);
+                    removeWorkerGroup(nodeAddress);
+                    addDeadServer(Constants.WORKER_TYPE, nodeAddress);
+                    removeNodeData(nodeAddress);
+                    log.warn("Worker server {} connect to raft cluster timeout, last heartbeat {}, timeout config {} ms",
+                            nodeAddress, convertTimeToString(entry.getValue()), expireTime);
+                }
+            }
+        }
+
+        private void triggerListener(ConnectionState connectionState) {
+            connectionListeners.forEach(listener -> listener.onUpdate(connectionState));
+        }
+    }
+
+    private class SubscribeCheckTask implements Runnable {
+
+        private final Map<String, String> nodeDataMap = new ConcurrentHashMap<>();
+
+        @Override
+        public void run() {
+            subscribeCheck();
+        }
+
+        private void subscribeCheck() {
+            if (dataSubScribeMap.isEmpty()) {
+                return;
+            }
+            final Map<String, String> currentNodeDataMap = getNodeDataMap();
+            // find the different
+            Map<String, String> addedData = new HashMap<>();
+            Map<String, String> deletedData = new HashMap<>();
+            Map<String, String> updatedData = new HashMap<>();
+            for (Map.Entry<String, String> entry : currentNodeDataMap.entrySet()) {
+                final String oldData = nodeDataMap.get(entry.getKey());
+                if (oldData == null) {
+                    addedData.put(entry.getKey(), entry.getValue());
+                } else {
+                    HeartBeat newHeartBeat = HeartBeat.decodeHeartBeat(entry.getValue());
```

Comment:
I think registry should not care about the content, it should has its own judgment, which is independent of the node data content. WDYT?

Reply:
Like zookeeper, subscribe listener compare node hearbeat to notify listener.